src/host-config/proxy.ts

```diff
@@ -0,0 +1,270 @@
+import { promises as fs } from 'fs';
+import path from 'path';
+import { isRight } from 'fp-ts/lib/Either';
+import Reporter from 'io-ts-reporters';
+
+import type { RedsocksConfig, HostProxyConfig } from './types';
+import { ProxyConfig } from './types';
+import { pathOnBoot, readFromBoot, writeToBoot } from '../lib/host-utils';
+import { unlinkAll, mkdirp } from '../lib/fs-utils';
+import { isENOENT } from '../lib/errors';
+import log from '../lib/supervisor-console';
+import * as dbus from '../lib/dbus';
+
+const proxyBasePath = pathOnBoot('system-proxy');
+const noProxyPath = path.join(proxyBasePath, 'no_proxy');
+const redsocksConfPath = path.join(proxyBasePath, 'redsocks.conf');
+
+export const disallowedProxyFields = ['local_ip', 'local_port'];
+export const validRedsocksProxyTypes = [
+	'socks4',
+	'socks5',
+	'http-connect',
+	'http-relay',
+];
+
+const isAuthField = (field: string): boolean =>
+	['login', 'password'].includes(field);
+
+// ? is a lazy operator, so only the contents up until the first `}(?=\s|$)` is matched.
+// (?=\s|$) indicates that `}` must be followed by a whitespace or end of file to match,
+// in case there are user fields with brackets such as login or password fields.
+const blockRegexFor = (blockLabel: string) =>
+	new RegExp(`${blockLabel}\\s?{([\\s\\S]+?)}(?=\\s|$)`);
+
+const baseBlock = {
+	log_debug: 'off',
+	log_info: 'on',
+	log: 'stderr',
+	daemon: 'off',
+	redirector: 'iptables',
+};
+
+export class RedsocksConf {
+	public static stringify(config: RedsocksConfig): string {
+		const blocks: string[] = [];
+
+		if (config.redsocks) {
+			blocks.push(RedsocksConf.stringifyBlock('base', baseBlock));
+			blocks.push(
+				RedsocksConf.stringifyBlock('redsocks', config.redsocks, {
+					local_ip: '127.0.0.1',
+					local_port: 12345,
+				}),
+			);
```

Comment:
Why not 
```suggestion
			blocks.push(
				RedsocksConf.stringifyBlock('redsocks', {...config.redsocks, 
					local_ip: '127.0.0.1',
					local_port: 12345,
				}),
			);
```

Reply:
Adopted this suggestion [here](https://github.com/balena-os/balena-supervisor/pull/2329/commits/70b037651d2a4a8ba6c86aa14199036b3d470e5c#diff-87da4178ce07f20bae291890a95e15b8bf7990d89aeb29ec6bfe1f77b508c3e5R45-R49), good call